File: src/coding_agent/tools.py
import inspect
import os
import pathlib
import shlex
import subprocess

import requests
from bs4 import BeautifulSoup
from ddgs import DDGS

COMMAND_TIMEOUT = 120


def map_python_type_to_json(py_type):
    """Converts Python type hints to JSON schema types."""
    mapping = {
        str: "string",
        int: "integer",
        float: "number",
        bool: "boolean",
        list: "array",
        dict: "object",
    }
    # Default to string if no type hint is provided
    return mapping.get(py_type, "string")


def generate_schema(func):
    """Automatically generates an OpenAI/Claude tool JSON schema from a Python function."""
    sig = inspect.signature(func)

    properties = {}
    required = []

    # Loop through the function's arguments (skip internal-only params like callbacks)
    for name, param in sig.parameters.items():
        if name.endswith("_callback"):
            continue
        # Map the Python type hint (e.g., str) to JSON type (e.g., "string")
        param_type = map_python_type_to_json(param.annotation)

        properties[name] = {"type": param_type, "description": f"The {name} parameter."}

        # If the argument doesn't have a default value, it's required
        if param.default == inspect.Parameter.empty:
            required.append(name)

    return {
        "type": "function",
        "function": {
            "name": func.__name__,
            # Uses the function's docstring as the description for the LLM
            "description": inspect.getdoc(func) or f"Execute {func.__name__}",
            "parameters": {
                "type": "object",
                "properties": properties,
                "required": required,
            },
        },
    }


def run_terminal_command(command: str, output_callback=None) -> str:
    """Run a terminal command and return its output. Only accepts a single 'command' argument (the shell command string)."""

    try:
        process = subprocess.Popen(
            command,
            shell=True,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
        )
        output_lines = []
        assert process.stdout is not None
        for line in process.stdout:
            output_lines.append(line)
            if output_callback:
                output_callback(line)
        process.wait(timeout=COMMAND_TIMEOUT)
        assert process.stderr is not None
        stderr = process.stderr.read()
        if process.returncode == 0:
            return (
                "".join(output_lines).strip()
                or "Command executed successfully with no output."
            )
        else:
            return f"Error: {stderr.strip()}"
    except subprocess.TimeoutExpired:
        process.kill()
        return f"Error: The command timed out after {COMMAND_TIMEOUT} seconds."
    except Exception as e:
        return f"Execution Exception: {str(e)}"


def write_file(file_path: str, content: str) -> str:
    """Write content to a file. Parameters: 'file_path' (string, the file location to write to) and 'content' (string, the complete file content). Returns success message or error."""
    try:
        parent = os.path.dirname(file_path)
        if parent:
            os.makedirs(parent, exist_ok=True)
        with open(file_path, "w") as f:
            f.write(content)
        return f"Success! File saved to {file_path}."
    except Exception as e:
        return f"Error writing file: {str(e)}"


def web_search(query: str) -> str:
    """Perform a web search using DuckDuckGo and return the top 5 results.
    Returns the Title, URL, and a short snippet for each result.
    Use the 'read_webpage' tool on the provided URLs to read the full content."""

    try:
        with DDGS() as ddgs:
            # We limit to 5 results to keep the context window clean and focused
            results = list(ddgs.text(query, max_results=5))

        if not results:
            return "No results found. Try modifying your search query."

        formatted_results = []
        for i, res in enumerate(results, 1):
            formatted_results.append(
                f"Result {i}:\n"
                f"Title: {res.get('title')}\n"
                f"URL: {res.get('href')}\n"
                f"Snippet: {res.get('body')}\n"
            )

        return "\n".join(formatted_results)
    except Exception as e:
        return f"Search failed: {str(e)}"


def read_webpage(url: str) -> str:
    """Fetches a URL, strips away all HTML/CSS/JS, and returns clean text for the agent to read."""

    try:
        # 1. Fetch the raw HTML (Timeout is crucial so the agent doesn't hang!)
        response = requests.get(url, timeout=10)
        response.raise_for_status()

        # 2. Parse it and strip out the junk
        soup = BeautifulSoup(response.text, "html.parser")

        # Destroy script and style elements completely
        for script in soup(["script", "style", "nav", "footer", "header"]):
            script.extract()

        # 3. Get the clean text
        text = soup.get_text(separator=" ", strip=True)

        # Limit the text length just in case it's a massive terms of service page
        # ~15,000 characters is a safe chunk for an LLM to read quickly
        return text[:15000]
    except requests.exceptions.Timeout:
        return "Error: The webpage took too long to load."
    except Exception as e:
        return f"Error reading webpage: {str(e)}"


def list_directory(path: str = ".") -> str:
    """Lists the contents of a directory. Only accepts 'path' (directory path, defaults to '.'). Useful for exploring the file system."""

    try:
        items = os.listdir(path)
        # Filter out noisy directories that confuse the LLM
        clean_items = [
            item for item in items if not item.startswith(".") and item != "__pycache__"
        ]

        if not clean_items:
            return f"Directory '{path}' is empty."

        return f"Contents of {path}:\n" + "\n".join(f"- {item}" for item in clean_items)
    except FileNotFoundError:
        return f"Error: Directory '{path}' not found."
    except Exception as e:
        return f"Error reading directory: {str(e)}"


def read_file(file_path: str) -> str:
    """Reads a file and returns its contents with line numbers. Parameter: 'file_path' (string, the file location to read). Use this to inspect code."""

    try:
        with open(file_path, "r") as f:
            lines = f.readlines()

        # Context Window Protection: Prevent reading massive files
        if len(lines) > 2000:
            return f"Error: File is too large ({len(lines)} lines). I can only read up to 2000 lines to protect memory."

        # Add line numbers to the output (e.g., "1 | import os")
        numbered_content = ""
        for i, line in enumerate(lines, 1):
            numbered_content += f"{i:4d} | {line}"

        return numbered_content
    except FileNotFoundError:
        return f"Error: File '{file_path}' not found. Try using list_directory to see what files exist."
    except UnicodeDecodeError:
        return "Error: This looks like a binary file (like an image or PDF). I can only read text files."
    except Exception as e:
        return f"Error reading file: {str(e)}"


def replace_text_in_file(file_path: str, old_text: str, new_text: str) -> str:
    """Replaces old_text with new_text in the specified file. Parameters: 'file_path' (string, the file location), 'old_text' (string, exact text to find), and 'new_text' (string, replacement text). Returns success message or error."""

    try:
        with open(file_path, "r") as f:
            content = f.read()

        if old_text not in content:
            return (
                f"Error: The exact string '{old_text}' was not found in the file. "
                "Check your spelling, whitespace, and indentation, and try again."
            )

        updated_content = content.replace(old_text, new_text)

        with open(file_path, "w") as f:
            f.write(updated_content)

        return f"Success! Replaced text in {file_path}."
    except FileNotFoundError:
        return f"Error: File '{file_path}' not found."
    except Exception as e:
        return f"Error replacing text in file: {str(e)}"


SEARCH_SKIP_DIRS = {
    ".git",
    "node_modules",
    "__pycache__",
    ".venv",
    "venv",
    ".tox",
    ".mypy_cache",
    ".ruff_cache",
}
MAX_FILE_SIZE = 1_000_000  # 1 MB


def search_files(
    pattern: str, path: str = ".", include: str = "", max_results: int = 20
) -> str:
    """Search for files by name glob or search file contents by text. If pattern contains '*' or '?' it performs a filename glob search (e.g. '**/*.py'). Otherwise it searches file contents for the pattern as a case-insensitive substring. Use 'include' to filter by extensions (e.g. '.py,.js'). Returns matching file paths with line numbers and context."""

    root = pathlib.Path(path).resolve()
    if not root.is_dir():
        return f"Error: Directory '{path}' not found."

    is_glob = "*" in pattern or "?" in pattern

    if is_glob:
        return _glob_search(root, pattern, max_results)
    else:
        extensions = (
            [e.strip() for e in include.split(",") if e.strip()] if include else []
        )
        return _content_search(root, pattern, extensions, max_results)


def _should_skip(path: pathlib.Path) -> bool:
    """Check if any path component is in the skip list."""
    return any(part in SEARCH_SKIP_DIRS for part in path.parts)


def _glob_search(root: pathlib.Path, pattern: str, max_results: int) -> str:
    """Find files matching a glob pattern."""
    matches = []
    for match in root.glob(pattern):
        if _should_skip(match) or not match.is_file():
            continue
        try:
            rel = match.relative_to(root)
        except ValueError:
            rel = match
        matches.append(str(rel))
        if len(matches) >= max_results:
            break

    if not matches:
        return f"No files matched the pattern '{pattern}' in {root}."

    result = f"Found {len(matches)} file(s) matching '{pattern}':\n"
    result += "\n".join(f"  {m}" for m in sorted(matches))
    if len(matches) == max_results:
        result += f"\n  (limited to {max_results} results)"
    return result


def _content_search(
    root: pathlib.Path, pattern: str, extensions: list, max_results: int
) -> str:
    """Search file contents for a substring, returning matches with context."""
    pattern_lower = pattern.lower()
    matches = []
    files_searched = 0

    for filepath in root.rglob("*"):
        if not filepath.is_file() or _should_skip(filepath):
            continue
        if extensions and filepath.suffix not in extensions:
            continue
        try:
            size = filepath.stat().st_size
        except OSError:
            continue
        if size > MAX_FILE_SIZE or size == 0:
            continue

        try:
            content = filepath.read_text(errors="ignore")
        except (OSError, UnicodeDecodeError):
            continue

        files_searched += 1
        lines = content.splitlines()
        for i, line in enumerate(lines):
            if pattern_lower in line.lower():
                try:
                    rel = filepath.relative_to(root)
                except ValueError:
                    rel = filepath

                context_lines = []
                start = max(0, i - 2)
                end = min(len(lines), i + 3)
                for j in range(start, end):
                    marker = ">" if j == i else " "
                    context_lines.append(f"  {marker} {j + 1:4d} | {lines[j]}")

                matches.append(
                    {
                        "file": str(rel),
                        "line": i + 1,
                        "context": "\n".join(context_lines),
                    }
                )

                if len(matches) >= max_results:
                    break
        if len(matches) >= max_results:
            break

    if not matches:
        return f"No matches for '{pattern}' in {files_searched} files searched."

    result = f"Found {len(matches)} match(es) for '{pattern}':\n\n"
    for m in matches:
        result += f"{m['file']}:{m['line']}\n{m['context']}\n\n"
    if len(matches) == max_results:
        result += f"(limited to {max_results} results)"
    return result


def run_git_command(command: str) -> str:
    """Run a git command and return its output. Useful for version control tasks."""

    if ";" in command or "&&" in command or "|" in command:
        return "Error: Command chaining is not allowed for security reasons. Run one git command at a time."

    command = command.strip()
    if command.startswith("git "):
        command = command[4:]

    try:
        args = ["git"] + shlex.split(command)

        result = subprocess.run(
            args, capture_output=True, text=True, timeout=COMMAND_TIMEOUT
        )
        if result.returncode == 0:
            return (
                result.stdout.strip()
                or "Git command executed successfully with no output."
            )
        else:
            return f"Error: {result.stderr.strip()}"
    except subprocess.TimeoutExpired:
        return f"Error: The git command timed out after {COMMAND_TIMEOUT} seconds."
    except Exception as e:
        return f"Execution Exception: {str(e)}"


# A router dictionary so the Agent can easily trigger the right function
AVAILABLE_TOOLS = {
    "run_terminal_command": run_terminal_command,
    "write_file": write_file,
    "web_search": web_search,
    "read_webpage": read_webpage,
    "list_directory": list_directory,
    "read_file": read_file,
    "replace_text_in_file": replace_text_in_file,
    "search_files": search_files,
    "run_git_command": run_git_command,
}

TOOL_SCHEMAS = [generate_schema(func) for func in AVAILABLE_TOOLS.values()]

File: coding-agent-rust/src/tools.rs
use serde_json::{json, Value};
use std::fs;
use std::path::Path;
use std::process::Command;
use std::time::Duration;
use tokio::time::timeout;
use scraper::Html;

const DANGEROUS_KEYWORDS: &[&str] = &["rm", "sudo", "dd", "mkfs", "format", "shutdown", "reboot", "poweroff"];

/// Get the OpenAI-format tool schemas
pub fn get_tools_schema() -> Value {
    json!([
        {
            "type": "function",
            "function": {
                "name": "run_terminal_command",
                "description": "Run a shell command with 15s timeout. Checks for dangerous keywords and prompts for confirmation if found.",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "command": {
                            "type": "string",
                            "description": "The shell command to execute"
                        }
                    },
                    "required": ["command"]
                }
            }
        },
        {
            "type": "function",
            "function": {
                "name": "write_file",
                "description": "Create parent directories if needed and write content to a file",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "file_path": {
                            "type": "string",
                            "description": "The path where to write the file"
                        },
                        "content": {
                            "type": "string",
                            "description": "The content to write to the file"
                        }
                    },
                    "required": ["file_path", "content"]
                }
            }
        },
        {
            "type": "function",
            "function": {
                "name": "read_file",
                "description": "Read a file and return its contents with line numbers",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "file_path": {
                            "type": "string",
                            "description": "The path to the file to read"
                        }
                    },
                    "required": ["file_path"]
                }
            }
        },
        {
            "type": "function",
            "function": {
                "name": "replace_text_in_file",
                "description": "Replace all occurrences of old_text with new_text in a file",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "file_path": {
                            "type": "string",
                            "description": "The path to the file"
                        },
                        "old_text": {
                            "type": "string",
                            "description": "The text to find and replace"
                        },
                        "new_text": {
                            "type": "string",
                            "description": "The replacement text"
                        }
                    },
                    "required": ["file_path", "old_text", "new_text"]
                }
            }
        },
        {
            "type": "function",
            "function": {
                "name": "list_directory",
                "description": "List directory contents, filtering out dotfiles and __pycache__",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "path": {
                            "type": "string",
                            "description": "The directory path (default: '.')"
                        }
                    }
                }
            }
        },
        {
            "type": "function",
            "function": {
                "name": "web_search",
                "description": "Search DuckDuckGo for the given query and return top 5 results",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "query": {
                            "type": "string",
                            "description": "The search query"
                        }
                    },
                    "required": ["query"]
                }
            }
        },
        {
            "type": "function",
            "function": {
                "name": "read_webpage",
                "description": "Fetch a webpage, parse HTML, and return clean text (up to 15000 chars)",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "url": {
                            "type": "string",
                            "description": "The URL to fetch"
                        }
                    },
                    "required": ["url"]
                }
            }
        },
        {
            "type": "function",
            "function": {
                "name": "run_git_command",
                "description": "Run a git command with validation and 15s timeout",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "command": {
                            "type": "string",
                            "description": "The git command (with or without 'git ' prefix)"
                        }
                    },
                    "required": ["command"]
                }
            }
        }
    ])
}

/// Execute a tool with the given name and arguments
pub async fn execute_tool(name: &str, args: &Value) -> String {
    match name {
        "run_terminal_command" => run_terminal_command(args).await,
        "write_file" => write_file(args),
        "read_file" => read_file(args),
        "replace_text_in_file" => replace_text_in_file(args),
        "list_directory" => list_directory(args),
        "web_search" => web_search(args).await,
        "read_webpage" => read_webpage(args).await,
        "run_git_command" => run_git_command(args).await,
        _ => format!("Unknown tool: {}", name),
    }
}

async fn run_terminal_command(args: &Value) -> String {
    let command = match args.get("command").and_then(|v| v.as_str()) {
        Some(cmd) => cmd,
        None => return "Error: 'command' parameter is required".to_string(),
    };

    // Check for dangerous keywords
    for keyword in DANGEROUS_KEYWORDS {
        if command.contains(keyword) {
            println!("\n⚠️  Warning: Command contains '{}' which is potentially dangerous.", keyword);
            println!("Command: {}", command);
            print!("Proceed? (y/N): ");
            std::io::Write::flush(&mut std::io::stdout()).ok();

            let mut input = String::new();
            if std::io::stdin().read_line(&mut input).is_err() {
                return "Command cancelled by user".to_string();
            }

            if !input.trim().eq_ignore_ascii_case("y") {
                return "Command cancelled by user".to_string();
            }
        }
    }

    // Execute with timeout
    let output = tokio::task::spawn_blocking({
        let command = command.to_string();
        move || {
            Command::new("sh")
                .arg("-c")
                .arg(&command)
                .output()
        }
    });

    match timeout(Duration::from_secs(15), output).await {
        Ok(Ok(output)) => {
            if output.status.success() {
                String::from_utf8_lossy(&output.stdout).to_string()
            } else {
                String::from_utf8_lossy(&output.stderr).to_string()
            }
        }
        Ok(Err(e)) => format!("Error spawning command: {}", e),
        Err(_) => "Command timed out after 15 seconds".to_string(),
    }
}

fn write_file(args: &Value) -> String {
    let file_path = match args.get("file_path").and_then(|v| v.as_str()) {
        Some(path) => path,
        None => return "Error: 'file_path' parameter is required".to_string(),
    };

    let content = match args.get("content").and_then(|v| v.as_str()) {
        Some(content) => content,
        None => return "Error: 'content' parameter is required".to_string(),
    };

    // Create parent directories
    if let Some(parent) = Path::new(file_path).parent() {
        if !parent.as_os_str().is_empty() {
            if let Err(e) = fs::create_dir_all(parent) {
                return format!("Error creating directories: {}", e);
            }
        }
    }

    // Write file
    match fs::write(file_path, content) {
        Ok(_) => format!("File written successfully to {}", file_path),
        Err(e) => format!("Error writing file: {}", e),
    }
}

fn read_file(args: &Value) -> String {
    let file_path = match args.get("file_path").and_then(|v| v.as_str()) {
        Some(path) => path,
        None => return "Error: 'file_path' parameter is required".to_string(),
    };

    let content = match fs::read_to_string(file_path) {
        Ok(content) => content,
        Err(e) => return format!("Error reading file: {}", e),
    };

    let lines: Vec<&str> = content.lines().collect();
    if lines.len() > 2000 {
        return format!(
            "Error: File has {} lines, exceeds 2000 line limit",
            lines.len()
        );
    }

    let formatted: Vec<String> = lines
        .iter()
        .enumerate()
        .map(|(i, line)| format!("{:4} | {}", i + 1, line))
        .collect();

    formatted.join("\n")
}

fn replace_text_in_file(args: &Value) -> String {
    let file_path = match args.get("file_path").and_then(|v| v.as_str()) {
        Some(path) => path,
        None => return "Error: 'file_path' parameter is required".to_string(),
    };

    let old_text = match args.get("old_text").and_then(|v| v.as_str()) {
        Some(text) => text,
        None => return "Error: 'old_text' parameter is required".to_string(),
    };

    let new_text = match args.get("new_text").and_then(|v| v.as_str()) {
        Some(text) => text,
        None => return "Error: 'new_text' parameter is required".to_string(),
    };

    let mut content = match fs::read_to_string(file_path) {
        Ok(content) => content,
        Err(e) => return format!("Error reading file: {}", e),
    };

    if !content.contains(old_text) {
        return "Error: old_text not found in file".to_string();
    }

    content = content.replace(old_text, new_text);

    match fs::write(file_path, content) {
        Ok(_) => "File updated successfully".to_string(),
        Err(e) => format!("Error writing file: {}", e),
    }
}

fn list_directory(args: &Value) -> String {
    let path = args
        .get("path")
        .and_then(|v| v.as_str())
        .unwrap_or(".");

    match fs::read_dir(path) {
        Ok(entries) => {
            let mut items: Vec<String> = entries
                .filter_map(|entry| {
                    entry.ok().and_then(|e| {
                        let name = e.file_name();
                        let name_str = name.to_string_lossy();

                        // Filter out dotfiles and __pycache__
                        if name_str.starts_with('.') || name_str == "__pycache__" {
                            return None;
                        }

                        Some(name_str.to_string())
                    })
                })
                .collect();

            items.sort();
            items.join("\n")
        }
        Err(e) => format!("Error reading directory: {}", e),
    }
}

async fn web_search(args: &Value) -> String {
    let query = match args.get("query").and_then(|v| v.as_str()) {
        Some(q) => q,
        None => return "Error: 'query' parameter is required".to_string(),
    };

    let encoded_query = urlencoding::encode(query);
    let url = format!("https://html.duckduckgo.com/html/?q={}", encoded_query);

    let client = reqwest::Client::new();
    match client.get(&url).send().await {
        Ok(response) => match response.text().await {
            Ok(html) => parse_duckduckgo_results(&html),
            Err(e) => format!("Error reading response: {}", e),
        },
        Err(e) => format!("Error making request: {}", e),
    }
}

fn parse_duckduckgo_results(html: &str) -> String {
    let document = Html::parse_document(html);
    let mut results = Vec::new();

    // DuckDuckGo results are in divs with class "result__body"
    let selector = match scraper::Selector::parse("div.result") {
        Ok(sel) => sel,
        Err(_) => {
            return "Error parsing search results".to_string();
        }
    };

    for (idx, element) in document.select(&selector).enumerate() {
        if idx >= 5 {
            break;
        }

        let title_selector = match scraper::Selector::parse("h2 > a") {
            Ok(sel) => sel,
            Err(_) => continue,
        };

        let snippet_selector = match scraper::Selector::parse(".result__snippet") {
            Ok(sel) => sel,
            Err(_) => continue,
        };

        let title = element
            .select(&title_selector)
            .next()
            .and_then(|el| el.value().attr("title"))
            .unwrap_or("");

        let url = element
            .select(&title_selector)
            .next()
            .and_then(|el| el.value().attr("href"))
            .unwrap_or("");

        let snippet = element
            .select(&snippet_selector)
            .next()
            .map(|el| el.inner_html())
            .unwrap_or_default();

        // Remove HTML tags from snippet
        let snippet_text = Html::parse_document(&snippet)
            .root_element()
            .text()
            .collect::<Vec<_>>()
            .join(" ");

        results.push(format!(
            "Title: {}\nURL: {}\nSnippet: {}\n",
            title, url, snippet_text
        ));
    }

    if results.is_empty() {
        "No search results found".to_string()
    } else {
        results.join("\n---\n")
    }
}

async fn read_webpage(args: &Value) -> String {
    let url = match args.get("url").and_then(|v| v.as_str()) {
        Some(u) => u,
        None => return "Error: 'url' parameter is required".to_string(),
    };

    let client = reqwest::Client::new();
    match timeout(
        Duration::from_secs(10),
        client.get(url).send(),
    )
    .await
    {
        Ok(Ok(response)) => match response.text().await {
            Ok(html) => {
                let document = Html::parse_document(&html);
                
                // Remove script, style, nav, footer, header elements
                let mut text_parts: Vec<String> = Vec::new();
                
                for element in document.root_element().descendants() {
                    if let scraper::node::Node::Element(e) = element.value() {
                        let name = e.name();
                        if name == "script" || name == "style" || name == "nav" 
                            || name == "footer" || name == "header" {
                            continue;
                        }
                    }
                }

                // Extract text
                let text = document
                    .root_element()
                    .text()
                    .collect::<Vec<_>>()
                    .join(" ");

                // Clean up whitespace
                let cleaned = text
                    .split_whitespace()
                    .collect::<Vec<_>>()
                    .join(" ");

                // Truncate to 15000 chars
                if cleaned.len() > 15000 {
                    cleaned[..15000].to_string()
                } else {
                    cleaned
                }
            }
            Err(e) => format!("Error reading response: {}", e),
        },
        Ok(Err(e)) => format!("Error making request: {}", e),
        Err(_) => "Request timed out after 10 seconds".to_string(),
    }
}

async fn run_git_command(args: &Value) -> String {
    let command = match args.get("command").and_then(|v| v.as_str()) {
        Some(cmd) => cmd,
        None => return "Error: 'command' parameter is required".to_string(),
    };

    // Check for shell metacharacters
    if command.contains(';') || command.contains("&&") || command.contains('|') {
        return "Error: Shell metacharacters (;, &&, |) are not allowed".to_string();
    }

    // Strip leading "git " if present
    let cmd_clean = if command.starts_with("git ") {
        &command[4..]
    } else {
        command
    };

    // Parse arguments
    let args_vec = match shell_words::split(cmd_clean) {
        Ok(args) => args,
        Err(e) => return format!("Error parsing command: {}", e),
    };

    if args_vec.is_empty() {
        return "Error: No git command provided".to_string();
    }

    // Execute git command
    let output = tokio::task::spawn_blocking({
        let args = args_vec.clone();
        move || {
            Command::new("git")
                .args(&args)
                .output()
        }
    });

    match timeout(Duration::from_secs(15), output).await {
        Ok(Ok(output)) => {
            if output.status.success() {
                String::from_utf8_lossy(&output.stdout).to_string()
            } else {
                String::from_utf8_lossy(&output.stderr).to_string()
            }
        }
        Ok(Err(e)) => format!("Error spawning command: {}", e),
        Err(_) => "Command timed out after 15 seconds".to_string(),
    }
}
